src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotIgnoreReturnValue.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    using static MicrosoftNetCoreAnalyzersResources;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotIgnoreReturnValueAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string CA2022RuleId = "CA2022";
+
+        internal static readonly DiagnosticDescriptor DoNotIgnoreReturnValueRule = DiagnosticDescriptorHelper.Create(
+            CA2022RuleId,
+            CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueTitle)),
+            CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueMessage)),
+            DiagnosticCategory.Reliability,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor DoNotIgnoreReturnValueRuleWithMessage = DiagnosticDescriptorHelper.Create(
+            CA2022RuleId,
+            CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueTitle)),
+            CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueMessageCustom)),
+            DiagnosticCategory.Reliability,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(DoNotIgnoreReturnValueDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(DoNotIgnoreReturnValueRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterOperationBlockStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemDiagnosticsCodeAnalysisDoNotIgnoreAttribute, out var doNotIgnoreAttribute))
```

Comment:
This is an attribute I think we expect individual assemblies might end up defining for themselves internally, e.g. System.Collections.Immutable in its downlevel builds will likely ship its own internal copy of this attribute. If such an assembly is then consumed by some other library, which then also sees the "real" attribute or its own, will this always do the right thing? I'm wondering if there are situations where it might find a different version and then will end up with false negatives later when comparing the attribute.

Reply:
My understanding is that it will do the right thing (based on how I've seen us identify attributes in other analyzers). As far as I know, it's only tied to the fully-qualified name and nothing else.